src/compiler/checker.ts

```diff
@@ -9187,7 +9187,8 @@ namespace ts {
                     }
                 }
             }
-            return createArrayType(elementTypes.length ? getUnionType(elementTypes) : emptyArrayElementType);
+            const elementType = getUnionType(elementTypes);
+            return createArrayType(elementType === neverType ? strictNullChecks ? wideningNeverType : undefinedType : elementType);

```

Comment:
Yes, the array literal would widen to `any[]`. We could do extra work to aggregate widening and only widen only if all (or if just one?) of the elements are of the widening `never` type. But it hardly seems worth it.